perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java

```diff
@@ -1366,23 +1364,14 @@ public void deleteApplication(PerunSession sess, Application app) throws PerunEx
 
 			if (AppState.NEW.equals(app.getState()) || AppState.REJECTED.equals(app.getState())) {
 
-				// Try to get reservedLogin and reservedNamespace before deletion
-				List<Pair<String, String>> logins;
-				try {
-					logins = jdbc.query("select namespace,login from application_reserved_logins where app_id=?", (resultSet, arg1) -> new Pair<>(resultSet.getString("namespace"), resultSet.getString("login")), app.getId());
-				} catch (EmptyResultDataAccessException e) {
-					// set empty logins
-					logins = new ArrayList<>();
-				}
-				// delete passwords in KDC
+				List<Pair<String, String>> logins = usersManager.getReservedLoginsOnlyByGivenApp(sess, app.getId());
```

Comment:
I wonder, if we shouldn't have some kind of synchronization, since we could be deleting various applications for the same user with same reserved login at once and it could be resolved like there is still some other application holding the reserved login and it would be kept.

Reply:
When I think about it more, we could probably do something like select for update when getting specific namespace/login combination from reserved logins table so we can be sure, that other processes are waiting for the first one to complete its thing.

Reply:
Good point, I added select for update to every select from the reserved logins table.

Reply:
Just a note, since we select login,namespace for the same user one by one, two processes could lock each other, but we don't generally use more than one login on registration form and probability of this is very low. We shall see in our manual tests.

Reply:
I think this is prevented by locking them in alphabetical order.